src/mip/utils/timestamp.hpp

```diff
@@ -0,0 +1,222 @@
+    ///     setWeek(timestamp, 4);
+    template<typename D>
+    void setWeek(D &timestamp, int week);
+
+    /// Sets a new time of week for the timestamp.
+    ///
+    /// The resulting time since epoch is calculated using the old week number and 
+    /// the new time of week.
+    ///
+    /// Example usage:
+    ///     // Setting time of week with the same unit duration as the timestamp.
+    ///     Seconds timestamp = TimestampState{}.getTimestamp<Seconds>();
+    ///     setTimeOfWeek<Seconds>(timestamp, 300);
+    ///     // ---> The timestamp will now be different! It will be the result from 
+    ///     //      setting the new time of week to 300 seconds. 
+    ///
+    ///     // Setting time of week with a different unit duration from the timestamp.
+    ///     Nanoseconds timestamp = TimestampState{}.getTimestamp<Nanoseconds>();
+    ///     setTimeOfWeek<Seconds>(timestamp, 6);
+    ///     // ---> The timestamp will now be the result from setting the new time of 
+    ///     //      week to 6 seconds (6,000,000,000 nanoseconds).
+    template<typename DTimeSet, typename DIn> 
+    void setTimeOfWeek(DIn &timestamp, int time);
+    
+    // TODO: Add manual timestamp update function.
+
+
+    /**************************************************************************************/
+    /* NOTE: The following are definitions for all template declarations above. There are */
+    /*       no new declarations following this statement.                                */
+    /**************************************************************************************/
+
+    template<typename D> 
+    inline D getTimeOfWeek(const D &timestamp)
+    {
+        if (timestamp <= Weeks{1}) {return timestamp;}
+
+        return timestamp - duration_cast<Weeks>(timestamp);
+    }
+
+    template<typename DCompare, typename D1, typename D2> 
+    inline bool timeChanged(const D1 &timestamp1, const D2 &timestamp2)
+    {
+        if (timestamp1 >= timestamp2)
+            return (timestamp1 - timestamp2 >= DCompare{1}) ? true : false;
+
+        return (timestamp2 - timestamp1 >= DCompare{1}) ? true : false;
+    }
+
+    template<typename T, typename DCast, typename DIn> 
+    inline T castTime(const DIn &timestamp)
+    {
+        return castTime<T>(duration_cast<DCast>(timestamp));
+    }
+
+    template<typename T, typename D> 
+    inline T castTime(const D &timestamp)
+    {
+        return static_cast<T>(timestamp.count());
+    }
+    
+    template<typename D>
+    inline void setWeek(D &timestamp, int week)
+    {
+        assert (week > 0); 
+        if (week <= 0) return;
+
+        timestamp = duration_cast<D>(Weeks(week) + getTimeOfWeek(timestamp));
+    }
+
+    template<typename DTimeSet, typename DIn> 
+    inline void setTimeOfWeek(DIn &timestamp, int time)
+    {
+        assert (time > 0);
+        if (time <= 0) return;
+
+        timestamp = duration_cast<DIn>(duration_cast<Weeks>(timestamp) + DTimeSet(time));
+    }
```

Comment:
No need to verify here. If `duration_cast<Weeks>(time) < 1`, then the week number is 0, so the new timestamp is just the time of week being set (because a full week hasn't elapsed yet). Though the week number starts at 1 (see above) anyway.

Reply:
We can't tell anything about the value of the `time` param unless we verify it?

What is the desired behavior for
```
int tow = 604805; // 1 week and 5 seconds
setTimeOfWeek<Seconds>(ts, tow);
```

Reply:
I see what you're saying now. Thought you meant for when the time had already been set, not for the incoming time to set. I will add the verification. Good catch!